src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -106,12 +110,15 @@ public int Count
             }
         }
 
-        public Enumerator GetEnumerator()
+        public T this[int index]
         {
-            if (_array == null)
-                ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
-            Contract.EndContractBlock();
+            get { return _array[_offset + index]; }
```

Comment:
@jkotas I am surprised, I thought you would have said no. I think it could discourage people from using the new API since `ArraySegment<T>` is mostly used in perf-sensitive places. However, I suppose if we throw an exception here we could always decide to allow it in the future.